contracts/ColonyNetworkStaking.sol

```diff
@@ -0,0 +1,306 @@
+    // it in ReputationMiningCycle.invalidateHash;
+    require(msg.sender == reputationMiningCycle);
+    for (uint256 i = 0; i < stakers.length; i++) {
+      // This is pretty harsh! Are we happy with this?
+      // Alternative: lose more than they would have gained for backing the right hash.
+      stakedBalances[stakers[i]] = 0;
+    }
+    // TODO: Where do these staked tokens go? Maybe split between the person who did the 'invalidate' transaction
+    // and the colony network?
+    // TODO: Lose rep?
+  }
+
+  function rewardStakers(address[] stakers) internal {
+    // Internal unlike punish, because it's only ever called from setReputationRootHash
+
+    // TODO: Actually think about this function
+    // Passing an array so that we don't incur the EtherRouter overhead for each staker if we looped over
+    // it in ReputationMiningCycle.invalidateHash;
+    address commonColonyAddress = _colonies["Common Colony"];
+    uint256 reward = 10**18; //TODO: Actually work out how much reputation they earn, based on activity elsewhere in the colony.
+    if (reward >= uint256(int256(-1))/2) {
+      reward = uint256(int256(-1))/2;
+    }
+    // TODO: We need to be able to prove that the assert on the next line will never happen, otherwise we're locked out of reputation mining.
+    // Something like the above cap is an adequate short-term solution, but at the very least need to double check the limits
+    // (which I've fingered-in-the-air, but could easily have an OBOE hiding inside).
+    assert(reward < uint256(int256(-1))); // We do a cast later, so make sure we don't overflow.
+    IColony(commonColonyAddress).mintTokensForColonyNetwork(stakers.length * reward); // This should be the total amount of new tokens we're awarding.
+    for (uint256 i = 0; i < stakers.length; i++) {
+      // We *know* we're the first entries in this reputation update log, so we don't need all the bookkeeping in
+      // the AppendReputationUpdateLog function
+      ReputationUpdateLog[activeReputationUpdateLog].push(ReputationLogEntry(
+        stakers[i], //The staker getting the reward
+        int256(reward),
+        0, //TODO: Work out what skill this should be. This should be a special 'mining' skill.
+        commonColonyAddress, // They earn this reputation in the common colony.
+        4, // Updates the user's skill, and the colony's skill, both globally and for the special 'mining' skill
+        i*4)//We're zero indexed, so this is the number of updates that came before in the reputation log.
+      );
+
+      // Also give them some newly minted tokens.
+      // We reinvest here as it's much easier (gas-wise).
+      stakedBalances[stakers[i]] += reward;
+    }
+  }
+}
+
+// TODO: Can we handle a dispute regarding the very first hash that should be set?
+
+
+contract ReputationMiningCycle {
+  address colonyNetworkAddress;
+  // TODO: Do we need both these mappings?
```

Comment:
I couldn't work out a way to do without them but maybe we can get rid of `submittedEntries` if we can accommodate `entry` in `submittedHashes` e.g.
```
struct Entry {
        address reputationMiner;
        uint N;
    }
mapping (bytes32 => mapping( uint256 => Entry[])) public submittedHashes;
```